clients/adapter-kafka/src/main/java/org/eclipse/hono/adapter/client/command/kafka/KafkaBasedInternalCommandConsumer.java

```diff
@@ -0,0 +1,237 @@
+        this.tracer = Objects.requireNonNull(tracer);
+    }
+
+    @Override
+    public Future<Void> start() {
+        // trigger creation of adapter specific topic and consumer
+        return createTopic().compose(v -> subscribeToTopic());
+    }
+
+    private Future<Void> createTopic() {
+        final Promise<Void> promise = Promise.promise();
+        final String topicName = getTopicName();
+        adminClient.createTopics(List.of(new NewTopic(topicName, NUM_PARTITIONS, REPLICATION_FACTOR)), promise.future());
+        return promise.future()
+                .onSuccess(v -> LOG.debug("created topic [{}]", topicName))
+                .onFailure(thr -> LOG.error("error creating topic [{}]", topicName, thr));
+    }
+
+    private Future<Void> subscribeToTopic() {
+        consumer.handler(this::handleCommandMessage);
+        consumer.exceptionHandler(thr -> {
+            LOG.error("consumer error occurred", thr);
+        });
+        consumer.partitionsRevokedHandler(partitionsSet -> {
+            LOG.debug("partitions revoked: [{}]", getPartitionsString(partitionsSet));
+        });
+        final Promise<Void> partitionAssignedPromise = Promise.promise();
+        consumer.partitionsAssignedHandler(partitionsSet -> {
+            LOG.debug("partitions assigned: [{}]", getPartitionsString(partitionsSet));
+            partitionAssignedPromise.tryComplete();
+        });
+        final String topicName = getTopicName();
+        final Promise<Void> subscribedPromise = Promise.promise();
+        consumer.subscribe(topicName, subscribedPromise);
+
+        return CompositeFuture.all(subscribedPromise.future(), partitionAssignedPromise.future())
+                .map((Void) null)
+                .onComplete(ar -> {
+                    // reset handler
+                    consumer.partitionsAssignedHandler(partitionsSet -> LOG.debug("partitions assigned: [{}]",
+                            getPartitionsString(partitionsSet)));
+                })
+                .onSuccess(v -> LOG.debug("subscribed and got partition assignment for topic [{}]", topicName))
+                .onFailure(thr -> LOG.error("error subscribing to topic [{}]", topicName, thr));
+    }
+
+    private String getPartitionsString(final Collection<TopicPartition> partitionsSet) {
+        return partitionsSet.stream()
+                .map(topicPartition -> Integer.toString(topicPartition.getPartition()))
+                .collect(Collectors.joining(", "));
+    }
+
+    private String getTopicName() {
+        return new HonoTopic(HonoTopic.Type.COMMAND_INTERNAL, adapterInstanceId).toString();
+    }
+
+    @Override
+    public Future<Void> stop() {
+        final String topicName = getTopicName();
+        final Promise<Void> adminClientClosePromise = Promise.promise();
+        LOG.debug("stop: delete topic [{}]", topicName);
+        adminClient.deleteTopics(List.of(topicName), ar -> {
```

Comment:
what if the process simply  crashes and thus the topics are not deleted during shut down of the component? Is there any other safeguard against *topic inflation* on the Kafka cluster?

Reply:
Cleaning up obsolete topics will be the task of the Command Router component. Should be prepared/implemented as part of https://github.com/eclipse/hono/issues/2028.

Reply:
I see